### PR TITLE
feat: unliquidator

### PR DESCRIPTION
**Pull request by dimsome**
Contract `Unliquidator.sol` for MIP 26. To stop liquidating stkAave and COMP and transfer to a set address (in this case Asset Management SubDao).
Done fork tests in `unliquidator.spec.ts` on Ethereum mainnet.

### CODE REVIEW

File: contracts/masset/liquidator/Unliquidator.sol [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * @notice Sets a new receive address for the tokens
+     * @param  _receiver  Address to which the tokens will be send at the end
+     */
+
+    function setReceiver(address _receiver) external onlyGovernance {
+        //
+        require(_receiver != address(0), "Invalid receiver address");
+        receiverSafe = _receiver;
```
Comment (doncesarts) [inline contracts/masset/liquidator/Unliquidator.sol:44]: Usually the same name is kept in order to have consistency . 
        receiverSafe = _receiver;

File: contracts/masset/liquidator/Unliquidator.sol [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+        emit NewReceiver(_receiver);
```
Comment (doncesarts) [inline contracts/masset/liquidator/Unliquidator.sol:46]: The Event ` NewReceiver`  gives the idea of a new receiver created . 
I suggest `ReceiverUpdated`

File: contracts/masset/liquidator/Unliquidator.sol [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    function triggerClaimAndDistribute(address _integration, address _token) external {
+        //
+        require(_integration != address(0), "Invalid integration address");
+        require(_token != address(0), "Invalid token address");
+
+        // 1. Claim rewards for the integration contract
+        IClaimRewards(_integration).claimRewards();
+
+        // 2. Send aave rewards to receiverSafe
```
Comment (doncesarts) [inline contracts/masset/liquidator/Unliquidator.sol:66]: The logic has been built agnostic to the tokens, rewards or integrator.  
The comments should be kept as such, 
`        // 2. Send token rewards to receiverSafe`

File: contracts/masset/liquidator/Unliquidator.sol [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    /***************************************
+                    CLAIM REWARDS
+    ****************************************/
+
+    /**
+     * @notice Claims the rewards and sends them to the receiverSafe, e.g. claims and sends stkAave
+     * @param  _integration  Integration address, this contract should have permissions to spend the token
+     * @param  _token  Address of the token that are claimed and send
+     */
+    function triggerClaimAndDistribute(address _integration, address _token) external {
```
Comment (doncesarts) [inline contracts/masset/liquidator/Unliquidator.sol:58]: Question: 
Why the verb `trigger` is used ? 
From my point of view `claimAndDistribute`  or `claimAndDistributeRewards` is more accurate.

File: contracts/masset/liquidator/Unliquidator.sol [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    function _sendRewards(address _from, address _token) internal {
+        //
+        // 1. Get balances of the token
+        uint256 amount = IERC20(_token).balanceOf(_from);
+        require(amount > 0, "No rewards to send");
+
+        // 2. Send the tokens to the receiverSafe
+        IERC20(_token).transferFrom(_from, receiverSafe, amount);
```
Comment (doncesarts) [inline contracts/masset/liquidator/Unliquidator.sol:89]: When calling `IERC20 .transferFrom ` the response of the method must be evaluated. 

```
 require(token.transferFrom(xxx,xxx,xxx), "Must sen tokens");
 ```
But  in this case that pattern is no longer used, mStable new contracts  use safeTransfer which comes from @openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol. T
hese need to add using SafeERC20 for IERC20; at the start of the contract.

File: contracts/masset/liquidator/Unliquidator.sol [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+        // 3. Emit the event
+        emit ClaimedAndSendRewards(_from, _token, amount, receiverSafe);
```
Comment (doncesarts) [inline contracts/masset/liquidator/Unliquidator.sol:92]: The event `ClaimedAndSendRewards`  is emit even if only the method `triggerDistribute`  is called.  
Therefore I am not sure if the name of the event is the correct of the place to trigger that  event is the right one.

**Comment by naddison36** [comment]
This looks good. So much easier when the tokens don't need to be liquidated

**Review by doncesarts — APPROVED** [review]

**Comment by doncesarts** [comment]
Hi @dimsome , in order to make all actions to pass, 
 - `coverage/coveralls — Coverage decreased (-0.02%) to 91.752%`

 I recommend you to :
1.    ignore the unliquidator  for test coverage at .solcover.js
2.   Open a github ticket, to increase the test coverage and mark it as `good first issue` 
3.  Merge this PR

**Pull request merged**